src/resources/pipe_win.cc

```diff
@@ -288,27 +333,86 @@ PRIMITIVE(fd_to_pipe) {
-      pipe_resource = _new WritePipeResource(resource_group, error, event);
-      break;
-    }
-    default:
-      FAIL(INVALID_ARGUMENT);
+  HANDLE handle = reinterpret_cast<HANDLE>(_get_osfhandle(fd));
+  if (handle == INVALID_HANDLE_VALUE) WINDOWS_ERROR;
+  int type = GetFileType(handle);
+  if (type != FILE_TYPE_PIPE && type != FILE_TYPE_CHAR) {
+    FAIL(INVALID_ARGUMENT);  // Not a pipe.
+  }
+
+  bool for_writing = fd != 0;  // Stdin vs stdout or stderr.
+
+  // If the pipe was in overlapped mode we could just make a PipeResource with
+  // _new WritePipeResource(resource_group, handle, event) or _new
+  // ReadPipeResource(resource_group, handle, event).  This is what our parent
+  // process has done if it is a Toit process.  But it is not normal to give a
+  // child process stdio pipes in overlapped mode, and it's really hard to
+  // detect it even if it happened (see
+  // https://microsoft.public.win32.programmer.kernel.narkive.com/VYscuhWn/was-handle-opened-using-file-flag-overlapped
+  // or https://archive.vn/wip/bmGhS), so we assume the pipes are in
+  // non-overlapped (synchronous) mode.
+
+  // Our pipe is not in overlapped mode, and unfortunately Windows has
+  // no way to switch to overlapped mode.  So we create a new pipe,
+  // and copy the data from the old pipe to the new pipe in a separate
+  // thread.
+  int read_overlap_flag = for_writing ? 0 : FILE_FLAG_OVERLAPPED;
+  int write_overlap_flag = for_writing ? FILE_FLAG_OVERLAPPED : 0;
+
+  char pipe_name_buffer[MAX_PATH];
+  snprintf(pipe_name_buffer,
+           MAX_PATH,
+           R"(\\.\Pipe\Toit.%08lx.%08lx)",
+           GetCurrentProcessId(),
+           InterlockedIncrement(&resource_group->pipe_serial_number())
+  );
+  HANDLE read = CreateNamedPipe(
+      pipe_name_buffer,
+      PIPE_ACCESS_INBOUND | read_overlap_flag,
+      PIPE_TYPE_BYTE | PIPE_WAIT,
+      1,             // Number of pipes.
+      8192,          // Out buffer size.
+      8192,          // In buffer size.
+      0,             // Default timeout (50 ms).
+      NULL           // Security attributes.
+  );
+  if (read == INVALID_HANDLE_VALUE) {
+    close_handle_keep_errno(event);
+    WINDOWS_ERROR;
+  }
+  HANDLE write = CreateFileA(
+      pipe_name_buffer,
+      GENERIC_WRITE,
+      0,                         // No sharing.
+      NULL,                      // Security attributes.
+      OPEN_EXISTING,
+      FILE_ATTRIBUTE_NORMAL | write_overlap_flag,
+      NULL                       // Template file.
+  );
+  if (write == INVALID_HANDLE_VALUE) {
+    close_handle_keep_errno(event);
+    close_handle_keep_errno(read);
+    WINDOWS_ERROR;
+  }
+  CopyPipeState* state = for_writing
+      ? _new CopyPipeState(read, handle)
+      : _new CopyPipeState(handle, write);
+  ASSERT(state);  // Can't fail on Windows.
+  HANDLE thread = CreateThread(NULL, 0, copy_pipe_thread, state, 0, NULL);
+  if (thread == NULL) {
+    close_handle_keep_errno(event);
```

Comment:
Is this the reason you don't want to call CloseHandle in a destructor on CopyPipeState? Can't you just use close_handle_keep_errno from within there?